accepted/build/build-publish-tweaks.md

```diff
@@ -0,0 +1,28 @@
+# Build and Publish tweaks in the .NET Core SDK
+
+In .NET Core 3, we are making Build output more similar to the Publish output.  In .NET Core 2.x and prior, dependencies from NuGet packages are not copied into the Build output.  At runtime, they are loaded from the NuGet package cache.  This meant that the Build output was not portable to other machines.
+
+For .NET Core 3, the Build output includes an app's dependencies.  The Build and Publish output from the base .NET SDK should be the same.  Additional operations that are not part of the inner-loop development cycle can be part of Publish, such as copying content for Web apps, trimming unused dependencies, or "linking" the app to create a single file executable.
+
+This document proposes some changes to the SDK behavior that will help us unify the Build and Publish logic.  Some of these could be breaking changes, but we believe the impact will be low.
+
+## `Publish="False"` on `PackageReference` should apply to Build
+
+If a `PackageReference` has `Publish="False"` metadata, then assets from that package will not be included in the Publish output.  In current previews of .NET Core 3, those assets will be included in the Build output but not the Publish output.  However, when targeting previous versions of .NET Core, no assets from NuGet packages would be included in the Build output.  Also, we are not aware of a scenario where you would want to specify that assets should not be in the Publish output, but would still want them in the Build output.
+
+So we propose that assets from a `PackageReference` with `Publish="False"` should be excluded from the Build output as well as the Publish outptu.
+
+## `PrivateAssets="All"` on `PackageReference` should not imply `Publish="False"`
+
+Setting `PrivateAssets="All"` on a `PackageReference` prevents that package from flowing to consuming projects.  This means that if a NuGet package is created from the project via Pack, the package with `PrivateAssets="All"` won't be included as a package dependency.  This is typically used for packages which are used as part of the build process of a library, but are not dependencies once the library has been compiled.  (See also "development dependency" packages.)
+
+However, `PrivateAssets="All"` by default also excludes assets from the package from the publish output.  It's not entirely clear why these two behaviors were linked to the same metadata, but it originates from the pre-1.0 project.json days.
+
+We think now is a good time to take the breaking change to separate these concepts.  The main use of PrivateAssets is to control whether assets flow to consuming projects, and the affect on Publish behavior doesn't seem to be documented.
```

Comment:
I think there's probably too much reliance on PrivateAssets=All -> no publish to make the break across the board. Can we base it on TFM?

Reply:
There are "build-time" only tools today that document putting `PrivateAssets=All` on your `PackageReference`. For example:

https://github.com/dotnet/sourcelink/blob/master/README.md#githubcom-and-github-enterprise

With this plan, these docs would need to change to also add `Publish=False`, correct?

The scenario is someone may use `Microsoft.SourceLink.GitHub` to get the commit hash embedded into their WPF/console/web application exe, but they don't want the SourceLink assemblies to get deployed with their application.

Reply:
I think it really depends on whether these packages have any actual runtime assets in lib or runtimes/. Otherwise, they'd technically be included in publish, but wouldn't actually contribute any assets to it, right? The particular package you mention has only build, buildCrossTargeting, and tools/ folders, so I would think not having Publish=false would be benign. Another one just like this that I'm aware of is xliff-tasks (because I wrote it) and is in the same boat. There, the crucial thing is not to flow the dependency. which private assets would still do. I'm not sure how to validate that most tooling packages of this kind are like this, though.

Reply:
Good point.

Just to clarify one part of your statement:

> I think it really depends on whether these packages have any actual runtime assets

Should be changed to

> I think it really depends on whether these packages _and their dependencies_ have any actual runtime assets

Reply:
Excellent clarification.